xla/service/gpu/ir_emission_utils.cc

```diff
@@ -740,6 +739,90 @@ bool IsIntermediate(const HloInstruction* instr, int allowed_operand_count) {
                                       instr->shape(), instr->dimensions()))));
 }
 
+bool IsReduceIntermediate(const HloInstruction* instr) {
+  if (instr->operand_count() > 1 || instr->user_count() > 1) {
+    return false;
+  }
+
+  // Only support elementwise ops that don't introduce additional compute.
+  // More benchmarking and better cost model are needed to enable this for
+  // more compute ops.
+  switch (instr->opcode()) {
+    case HloOpcode::kBitcast:
+    case HloOpcode::kBitcastConvert:
+    case HloOpcode::kConvert:
+      return true;
+    case HloOpcode::kReshape:
+      return ShapeUtil::ReshapeIsBitcast(instr->operand(0)->shape(),
+                                         instr->shape());
+    default:
+      return false;
+  }
+}
+
+// Recursive helper for GetFusionRoots below.
+static void GetFusionRootsRec(HloInstruction* root,
+                              std::vector<HloInstruction*>& out) {
+  if (root->opcode() == HloOpcode::kGetTupleElement) {
+    return GetFusionRootsRec(root->mutable_operand(0), out);
+  } else if (root->opcode() == HloOpcode::kTuple) {
+    for (int i = 0; i < root->operand_count(); i++) {
+      GetFusionRootsRec(root->mutable_operand(i), out);
+    }
+  } else {
+    if (!out.empty() && out.back() == root) {
+      return;
+    }
+    CHECK(!absl::c_linear_search(out, root))
+        << "Fusion root contains instruction " << root->ToString()
+        << " multiple times";
+    out.push_back(root);
+  }
+}
+
+std::vector<HloInstruction*> GetFusionRoots(HloComputation* computation) {
+  std::vector<HloInstruction*> out;
+  GetFusionRootsRec(computation->root_instruction(), out);
+  return out;
+}
+
+static const HloInstruction* FindNonTrivialReductionHero(
```

Comment:
Can this be moved to gpu_fusible? Also, I think non-const return and parameter would be better to avoid a const_cast later.

Reply:
yes, i will move all reduction hero related functions to reduction_utils

Reply:
You can also move it to reduction_utils if that works without reduction_utils depending on ir_emission_utils. But otherwise gpu_fusible would be a better place. Ideally I wanted reduction_utils and ir_emission_utils be independent from each other. The way I see this, these Hero helper functions are fusion related, so gpu_fusible is a good place for it.